rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+
+**TBD**: We would like to offer different plugin runtimes besides in-process as plugins themselves. These plugin runtimes would then come with their own dependencies that would be exposed to the parent service, such as gRPC or Lambda function invocation.
+
+## Drawbacks
+There are a handful major issues with the proposed direction:
+
+1. In-process plugins have a poor developer and operator experience which can negatively affect end-users, especially in rollback scenarios. In-process plugins also require developers to have more knowledge of the Spinnaker toolchain and potentially the service architecture. Exposing logging, telemetry and crash reports to plugin developers for in-process plugins will be difficult or impossible dependent on a Spinnaker installation’s preferences.
+2. Supporting alternative runtimes adds complexity to the final implementation and will likely require good judgment on an extension point developer to write the correct abstractions.
+
+## Prior Art and Alternatives
+https://github.com/spinnaker/spinnaker/issues/4181 … and linked PRs
+
+The existing plugins code that has come out from [#4181](https://github.com/spinnaker/spinnaker/issues/4181)  is a successful implementation of plugins, but only supports an in-process runtime model, and is implementing a bespoke plugin system. It would be preferential to use an existing library for plugins, where possible, that we don’t need to solely maintain.
+
+Furthermore, the existing system does not afford alternate plugin runtimes. In-process plugins have the benefit of being capable of a wide variety of changes, but come at the cost of a poor operator story in that service and plugin release cadences must be in lock-step. This model is fine for Spinnaker installations where the operator is also the plugin developer, but offers a poor story for environments where the plugin developer and Spinnaker service operators are not the same group and do not want to coordinate and be coupled to each other’s release cadences.
+
+The proposed plugin system will aim for feature parity of the existing plugin system, but will not guarantee backwards compatibility of developer or operator experience. The priority is in satisfying broad-strokes business requirements, as opposed to satisfying existing technical capabilities.
+
+## Known Unknowns
+There are a lot of known unknowns. It’s the intention that separate proposals will be made for many of these unknowns, such as development and publishing lifecycle, as well as operator experiences.
+
+* In-process plugins
+	* How will plugin authors get metrics, logs, or crash reports from their plugins?
+	* How will operators be notified that new plugin versions exist and can be released?
+	* Do we want to support updating backend plugins at runtime?
```

Comment:
not sure how to capture this... but say a plugin extends stages, the plugin was upgraded in a non-backwards compatible way but the messages in the queue/stage contexts are still in flight so the new plugin can't pickup where the old one left off. Obv, a burden on the plugin developer but we should make it clear in docs/maybe think of tck testing for this?

Reply:
I was thinking that plugins would need to be written in a backwards-compatible way. Plugin manifests will be able to define a minimum spinnaker version, but I had not addressed backwards-incompatible plugin behavior. This seems like something that falls on the plugin developer to manage, rather than something that we dictate and enforce via the core platform? WDYT?

Reply:
Yeah, I agree. I am just thinking how to we document/inform the developers that this is required/how do we present to the operators/end-users when a plugin fails vs. spinnaker fails. I think we should make sure that failure attribution is crystal clear

Reply:
That's a good thing to keep in mind. Sounds like a plugin developer documentation thing. Thank you!